src/lib/FreeC/Pass/TypeInferencePass.hs

```diff
@@ -666,6 +666,22 @@ annotateExprWith' (IR.Lambda srcSpan args expr _) resType =
     addTypeEquation srcSpan funcType resType
     return (IR.Lambda srcSpan args' expr' (makeExprType resType))
 
+
+annotateExprWith' (IR.Let srcSpan binds expr _) resType =
+  withLocalTypeAssumption $ do
+    retType  <- liftConverter freshTypeVar
+    binds'   <- mapM annotateBind binds
+    expr'    <- annotateExprWith expr retType
+    return (IR.Let srcSpan binds' expr' (makeExprType resType))
+     where
+       annotateBind :: IR.Bind -> TypeInference IR.Bind
+       annotateBind (IR.Bind bindSrcSpan bindVarPat bindExpr) = do
+           bindVarPat' <- annotateVarPat bindVarPat
```

Comment:
Since other bindings can refer to this binding but are entered by `annotateVarPat` into the type assumption, all variable patterns have to be annotated before the first binding.